[Site Isolation] Avoid reusing the main frame after cross-site navigation

charliewolfe opened the pull request:
#### 42e478a31abe6395c9b9133437f9fced3f112fff
<pre>
[Site Isolation] Avoid reusing the main frame after cross-site navigation
<a href="https://bugs.webkit.org/show_bug.cgi?id=285469">https://bugs.webkit.org/show_bug.cgi?id=285469</a>
<a href="https://rdar.apple.com/142438253">rdar://142438253</a>

Reviewed by Alex Christensen.

With site isolation, we were reusing the previous main frame even after swapping processes for the page.
This will cause issues when enabling the back/forward cache because, after swapping processes, we need to
put the previous main frame in a suspended state. I needed make several changes to make this work with
other site isolation changes that assumed we would always reuse the main frame. More details below.

* Source/WebKit/Shared/WebBackForwardListFrameItem.cpp:
(WebKit::WebBackForwardListFrameItem::sharesAncestor const):
(WebKit::WebBackForwardListFrameItem::hasAncestorFrame): Deleted.
* Source/WebKit/Shared/WebBackForwardListFrameItem.h:
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::ProvisionalPageProxy):
Keep reusing the main frame if it has an opener. We should probably not reuse the main frame in this case
either, but undoing that is more difficult because the opener needs to be able to keep a reference to the
main frame.

(WebKit::ProvisionalPageProxy::initializeWebPage):
(WebKit::ProvisionalPageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/WebBackForwardList.cpp:
(WebKit::WebBackForwardList::addItem):
hasAncestorFrame() assumed that frames that navigated would persist their frameID, which is no longer the
case after this change. Change the function to check if two items share any ancestor to keep previous
behavior.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::commitProvisionalPage):
With site isolation the process passed to ProvisionalPageProxy may be the current main frame process if
it’s in use by another frame.

Canonical link: <a href="https://commits.webkit.org/288547@main">https://commits.webkit.org/288547@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/74a6916a2bc4c235500d16eba8a0495a1aa28b27

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83604 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3221 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37904 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88675 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34612 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3309 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11179 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65034 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22776 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86650 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2433 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75964 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45321 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2344 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33660 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73413 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30914 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90052 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10869 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7845 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73467 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11092 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71787 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72692 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16939 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15647 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2193 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12932 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10821 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16293 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10669 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14144 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12441 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/WebBackForwardListFrameItem.cpp:
```diff
@@ -158,11 +158,13 @@ Ref<FrameState> WebBackForwardListFrameItem::copyFrameStateWithChildren()
-        if (ancestor->frameID() == frameID)
-            return true;
+    for (RefPtr currentAncestor = m_parent.get(); currentAncestor; currentAncestor = currentAncestor->m_parent.get()) {
+        for (RefPtr frameItemAncestor = frameItem.m_parent.get(); frameItemAncestor; frameItemAncestor = frameItemAncestor->m_parent.get()) {
```
This is a classic O(n^2) algorithm.  Could we instead make two sets of identifiers and check the intersection?  That would be O(n)

charliewolfe replied:
Sure

charliewolfe committed:
[Site Isolation] Avoid reusing the main frame after cross-site navigation
https://bugs.webkit.org/show_bug.cgi?id=285469
rdar://142438253

Reviewed by Alex Christensen.

With site isolation, we were reusing the previous main frame even after swapping processes for the page.
This will cause issues when enabling the back/forward cache because, after swapping processes, we need to
put the previous main frame in a suspended state. I needed make several changes to make this work with
other site isolation changes that assumed we would always reuse the main frame. More details below.

* Source/WebKit/Shared/WebBackForwardListFrameItem.cpp:
(WebKit::WebBackForwardListFrameItem::sharesAncestor const):
(WebKit::WebBackForwardListFrameItem::hasAncestorFrame): Deleted.
* Source/WebKit/Shared/WebBackForwardListFrameItem.h:
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::ProvisionalPageProxy):
Keep reusing the main frame if it has an opener. We should probably not reuse the main frame in this case
either, but undoing that is more difficult because the opener needs to be able to keep a reference to the
main frame.

(WebKit::ProvisionalPageProxy::initializeWebPage):
(WebKit::ProvisionalPageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/WebBackForwardList.cpp:
(WebKit::WebBackForwardList::addItem):
hasAncestorFrame() assumed that frames that navigated would persist their frameID, which is no longer the
case after this change. Change the function to check if two items share any ancestor to keep previous
behavior.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::commitProvisionalPage):
With site isolation the process passed to ProvisionalPageProxy may be the current main frame process if
it’s in use by another frame.

Canonical link: https://commits.webkit.org/288547@main
